File: airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/streaming/AdaptiveStreamingQueryConfig.java
/*
 * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.db.jdbc.streaming;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Optional;

public class AdaptiveStreamingQueryConfig implements JdbcStreamingQueryConfig {

  private final FetchSizeEstimator fetchSizeEstimator;

  public AdaptiveStreamingQueryConfig() {
    this.fetchSizeEstimator = TwoStageSizeEstimator.getInstance();
  }

  @Override
  public void initialize(final Connection connection, final Statement preparedStatement) throws SQLException {
    connection.setAutoCommit(false);
    preparedStatement.setFetchSize(FetchSizeConstants.INITIAL_SAMPLE_SIZE);
    LOGGER.info("Set initial fetch size: {}", preparedStatement.getFetchSize());
  }

  @Override
  public void accept(final ResultSet resultSet, final Object rowData) throws SQLException {
    fetchSizeEstimator.accept(rowData);
    final Optional<Integer> newFetchSize = fetchSizeEstimator.getFetchSize();
    if (newFetchSize.isPresent()) {
      resultSet.setFetchSize(newFetchSize.get());
      LOGGER.info("Updated fetch size: {}", resultSet.getFetchSize());
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/db/jdbc/streaming/AdaptiveStreamingQueryConfig.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.db.jdbc.streaming

import io.github.oshai.kotlinlogging.KotlinLogging
import java.sql.*

private val LOGGER = KotlinLogging.logger {}

open class AdaptiveStreamingQueryConfig : JdbcStreamingQueryConfig {
    private val fetchSizeEstimator: FetchSizeEstimator = TwoStageSizeEstimator.Companion.instance
    private var currentFetchSize: Int

    init {
        this.currentFetchSize = FetchSizeConstants.INITIAL_SAMPLE_SIZE
    }

    @Throws(SQLException::class)
    override fun initialize(connection: Connection, statement: Statement) {
        connection.autoCommit = false
        statement.fetchSize = FetchSizeConstants.INITIAL_SAMPLE_SIZE
        currentFetchSize = FetchSizeConstants.INITIAL_SAMPLE_SIZE
        LOGGER.info { "Set initial fetch size: ${statement.fetchSize} rows" }
    }

    @Throws(SQLException::class)
    override fun accept(resultSet: ResultSet, rowData: Any) {
        fetchSizeEstimator.accept(rowData)
        val newFetchSize = fetchSizeEstimator.fetchSize

        if (newFetchSize.isPresent && currentFetchSize != newFetchSize.get()) {
            LOGGER.info { "Set new fetch size: ${newFetchSize.get()} rows" }
            resultSet.fetchSize = newFetchSize.get()
            currentFetchSize = newFetchSize.get()
        }
    }
}
